[css-anchor-position-1] Implement swapping due to a try-tactic for self-alignment properties

tuankiet65 opened the pull request:
#### 79b8df526100e9ab5ae9fa45da73c00f595abdc0
<pre>
[css-anchor-position-1] Implement swapping due to a try-tactic for self-alignment properties
<a href="https://rdar.apple.com/148137397">rdar://148137397</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=290665">https://bugs.webkit.org/show_bug.cgi?id=290665</a>

Reviewed by Antti Koivisto.

This commit implements swapping due to a try-tactic [1] for self-alignment
properties (align-self and justify-self)

[1]: <a href="https://drafts.csswg.org/css-anchor-position-1/#swap-due-to-a-try-tactic">https://drafts.csswg.org/css-anchor-position-1/#swap-due-to-a-try-tactic</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/base-style-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-alignment-expected.txt:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::flipStart):
    - Swap values of align-self and justify-self when applying flip-start tactic.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::oppositeItemPosition):
    - Add function to flip ItemPosition values.

(WebCore::Style::BuilderConverter::convertSelfOrDefaultAlignmentData):
    - Flip position value when a position-try fallback is specified
      and the current property is align-self or justify-self.

Canonical link: <a href="https://commits.webkit.org/293328@main">https://commits.webkit.org/293328@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db88bf11d4d06d4ea6cd0ea2d2c7dcabb6101fa5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98585 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18218 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8454 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103706 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49147 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100629 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18511 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26669 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75045 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32198 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101589 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14040 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89034 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55403 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13819 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6999 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48555 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83775 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7077 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106079 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25675 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18701 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84018 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26052 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85235 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83503 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21095 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28145 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5825 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19358 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25633 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30814 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25451 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28771 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27026 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1919,6 +1961,46 @@ inline StyleSelfAlignmentData BuilderConverter::convertSelfOrDefaultAlignmentDat
     } else
         alignmentData.setPosition(fromCSSValue<ItemPosition>(value));
+
+    // Flip the position according to position-try fallback, if specified.
+    if (auto positionTryFallback = builderState.positionTryFallback()) {
```
We might want to factor applying positionTryFallback applying to a separate class (AnchorPositionEvaluator or a new one) as the amount of code related to it in the converter is getting pretty large.

tuankiet65 replied:
I suppose this would be the last (or close to the last) addition to fallback handling, so we could refactor later. Tracked here https://bugs.webkit.org/show_bug.cgi?id=291111

tuankiet65 committed:
[css-anchor-position-1] Implement swapping due to a try-tactic for self-alignment properties
rdar://148137397
https://bugs.webkit.org/show_bug.cgi?id=290665

Reviewed by Antti Koivisto.

This commit implements swapping due to a try-tactic [1] for self-alignment
properties (align-self and justify-self)

[1]: https://drafts.csswg.org/css-anchor-position-1/#swap-due-to-a-try-tactic

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/base-style-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-alignment-expected.txt:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::flipStart):
    - Swap values of align-self and justify-self when applying flip-start tactic.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::oppositeItemPosition):
    - Add function to flip ItemPosition values.

(WebCore::Style::BuilderConverter::convertSelfOrDefaultAlignmentData):
    - Flip position value when a position-try fallback is specified
      and the current property is align-self or justify-self.

Canonical link: https://commits.webkit.org/293328@main
